File: io/InStream.java
package common.io;

import common.CommonStatic;
import common.util.Data;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

public strictfp interface InStream {

	static InStream getIns(File f) {
		return Data.err(() -> new ISStream(f));
	}

	default byte[] nextBytesB() {
		int len = nextByte();
		byte[] ints = new byte[len];
		for (int i = 0; i < len; i++)
			ints[i] = (byte) nextByte();
		return ints;
	}

	default byte[] nextBytesI() {
		int len = nextInt();
		byte[] ints = new byte[len];
		for (int i = 0; i < len; i++)
			ints[i] = (byte) nextByte();
		return ints;
	}

	default double[] nextDoubles() {
		int len = nextByte();
		double[] ints = new double[len];
		for (int i = 0; i < len; i++)
			ints[i] = nextDouble();
		return ints;
	}

	default int[] nextIntsB() {
		int len = nextByte();
		int[] ints = new int[len];
		for (int i = 0; i < len; i++)
			ints[i] = nextInt();
		return ints;
	}

	default int[][] nextIntsBB() {
		int len = nextByte();
		int[][] ints = new int[len][];
		for (int i = 0; i < len; i++)
			ints[i] = nextIntsB();
		return ints;
	}

	default String nextString() {
		byte[] bts = nextBytesB();
		return new String(bts, StandardCharsets.UTF_8);
	}

	boolean end();

	int nextByte();

	double nextDouble();

	float nextFloat();

	int nextInt();

	long nextLong();

	int nextShort();

	InStream subStream();

	OutStream translate();

	default void close() throws IOException {

	}
}

strictfp class InStreamDef extends DataIO implements InStream {

	private final int[] bs;
	private final int off, max;
	private int index;

	InStreamDef(InStreamDef isd) {
		bs = isd.bs;
		index = off = isd.index;
		max = isd.max;
	}

	InStreamDef(int[] data, int ofs, int m) {
		bs = data;
		off = ofs;
		max = m;
		index = off;
	}

	@Override
	public boolean end() {
		return index == max;
	}

	@Override
	public int nextByte() {
		check(1);
		int ans = toByte(bs, index);
		index++;
		return ans;
	}

	@Override
	public double nextDouble() {
		check(8);
		double ans = toDouble(bs, index);
		index += 8;
		return ans;
	}

	@Override
	public float nextFloat() {
		check(4);
		float ans = toFloat(bs, index);
		index += 4;
		return ans;
	}

	@Override
	public int nextInt() {
		check(4);
		int ans = toInt(bs, index);
		index += 4;
		return ans;
	}

	@Override
	public long nextLong() {
		check(8);
		long ans = toLong(bs, index);
		index += 8;
		return ans;
	}

	@Override
	public int nextShort() {
		check(2);
		int ans = toShort(bs, index);
		index += 2;
		return ans;
	}

	public int pos() {
		return index - off;
	}

	public void reread() {
		index = off;
	}

	public int size() {
		return max - off;
	}

	public void skip(int n) {
		index += n;
	}

	@Override
	public InStreamDef subStream() {
		int n = nextInt();
		if (n > size()) {
			new Exception("error in getting subStream").printStackTrace();
			CommonStatic.def.save(false, true);
		}
		InStreamDef is = new InStreamDef(bs, index, index + n);
		index += n;
		return is;
	}

	@Override
	public String toString() {
		return "InStreamDef " + size();
	}

	@Override
	public OutStreamDef translate() {
		byte[] data = new byte[max - index];
		for (int i = 0; i < max - index; i++)
			data[i] = (byte) bs[index + i];
		return new OutStreamDef(data);
	}

	protected int[] getBytes() {
		return bs;
	}

	private void check(int i) {
		if (max - index < i) {
			String str = "out of bound: " + (index - off) + "/" + (max - off) + ", " + index + "/" + max + "/" + off
					+ "/" + bs.length;
			throw new BCUException(str);
		}
	}

}

File: io/InStream.kt
package common.io

import common.CommonStatic
import common.io.DataIO
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import java.io.UnsupportedEncodingException

@Strictfp
interface InStream {
    fun end(): Boolean
    fun nextByte(): Int
    fun nextBytesB(): ByteArray
    fun nextBytesI(): ByteArray
    fun nextDouble(): Double
    fun nextDoubles(): DoubleArray
    fun nextFloat(): Float
    fun nextInt(): Int
    fun nextIntsB(): IntArray
    fun nextIntsBB(): Array<IntArray?>
    fun nextLong(): Long
    fun nextShort(): Int
    fun nextString(): String
    fun subStream(): InStream
    fun translate(): OutStream

    companion object {
        fun getIns(bs: ByteArray): InStream? {
            val `is`: IntArray = DataIO.Companion.translate(bs)
            val sig: Int = DataIO.Companion.toInt(`is`, 0)
            if (sig == bs.size - 4) {
                return InStreamDef(`is`, 4, `is`.size)
            }
            throw BCUException("Unsupported version")
        }
    }
}

@Strictfp
internal class InStreamDef : DataIO, InStream {
    private val bs: IntArray
    private val off: Int
    private val max: Int
    private var index: Int

    protected constructor(isd: InStreamDef) {
        bs = isd.bs
        off = isd.index
        index = off
        max = isd.max
    }

    constructor(data: IntArray, ofs: Int, m: Int) {
        bs = data
        off = ofs
        max = m
        index = off
    }

    override fun end(): Boolean {
        return index == max
    }

    override fun nextByte(): Int {
        check(1)
        val ans: Int = DataIO.Companion.toByte(bs, index)
        index++
        return ans
    }

    override fun nextBytesB(): ByteArray {
        val len = nextByte()
        val ints = ByteArray(len)
        for (i in 0 until len) ints[i] = nextByte().toByte()
        return ints
    }

    override fun nextBytesI(): ByteArray {
        val len = nextInt()
        val ints = ByteArray(len)
        for (i in 0 until len) ints[i] = nextByte().toByte()
        return ints
    }

    override fun nextDouble(): Double {
        check(8)
        val ans: Double = DataIO.Companion.toDouble(bs, index)
        index += 8
        return ans
    }

    override fun nextDoubles(): DoubleArray {
        val len = nextByte()
        val ints = DoubleArray(len)
        for (i in 0 until len) ints[i] = nextDouble()
        return ints
    }

    override fun nextFloat(): Float {
        check(4)
        val ans: Float = DataIO.Companion.toFloat(bs, index)
        index += 4
        return ans
    }

    override fun nextInt(): Int {
        check(4)
        val ans: Int = DataIO.Companion.toInt(bs, index)
        index += 4
        return ans
    }

    override fun nextIntsB(): IntArray {
        val len = nextByte()
        val ints = IntArray(len)
        for (i in 0 until len) ints[i] = nextInt()
        return ints
    }

    override fun nextIntsBB(): Array<IntArray?> {
        val len = nextByte()
        val ints = arrayOfNulls<IntArray>(len)
        for (i in 0 until len) ints[i] = nextIntsB()
        return ints
    }

    override fun nextLong(): Long {
        check(8)
        val ans: Long = DataIO.Companion.toLong(bs, index)
        index += 8
        return ans
    }

    override fun nextShort(): Int {
        check(2)
        val ans: Int = DataIO.Companion.toShort(bs, index)
        index += 2
        return ans
    }

    override fun nextString(): String {
        val bts = nextBytesB()
        return try {
            String(bts, "UTF-8")
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
            String(bts)
        }
    }

    fun pos(): Int {
        return index - off
    }

    fun reread() {
        index = off
    }

    fun size(): Int {
        return max - off
    }

    fun skip(n: Int) {
        index += n
    }

    override fun subStream(): InStreamDef {
        val n = nextInt()
        if (n > size()) {
            Exception("error in getting subStream").printStackTrace()
            CommonStatic.def.exit(false)
        }
        val `is` = InStreamDef(bs, index, index + n)
        index += n
        return `is`
    }

    override fun toString(): String {
        return "InStreamDef " + size()
    }

    override fun translate(): OutStreamDef {
        val data = ByteArray(max - index)
        for (i in 0 until max - index) data[i] = bs[index + i].toByte()
        return OutStreamDef(data)
    }

    protected fun getBytes(): IntArray {
        return bs
    }

    private fun check(i: Int) {
        if (max - index < i) {
            val str = ("out of bound: " + (index - off) + "/" + (max - off) + ", " + index + "/" + max + "/" + off
                    + "/" + bs.size)
            throw BCUException(str)
        }
    }
}
